doc/source/train/doc_code/checkpoints.py

```diff
@@ -90,12 +89,19 @@ def train_func(config):
     optimizer = Adam(model.parameters(), lr=3e-4)
     criterion = nn.MSELoss()
 
+    # Wrap the model in DDP and move it to GPU.
+    model = ray.train.torch.prepare_model(model)
+
     # ====== Resume training state from the checkpoint. ======
     start_epoch = 0
     checkpoint = train.get_checkpoint()
     if checkpoint:
         with checkpoint.as_directory() as checkpoint_dir:
-            model.load_state_dict(torch.load(os.path.join(checkpoint_dir, "model.pt")))
+            model_state_dict = torch.load(
+                os.path.join(checkpoint_dir, "model.pt"),
+                # map_location=...,  # Load onto a different device if needed.
+            )
+            model.module.load_state_dict(model_state_dict)
```

Comment:
It's this issue: https://github.com/ray-project/ray/issues/40627

Model starts out on CPU, then gets moved to GPU with `prepare_model`. The model weights were saved from GPU and would be loaded back to GPU by default (unless `map_location`) is set, so there was a device mismatch issue before.

Then, because `prepare_model` also wraps in DDP, I need to do `model.module`. Alternatives are:
1. Save the weights without unwrapping the model (`model.state_dict()`), but then it's hard to use w/o ddp.
2. Map loaded weights to cpu, load into unwrapped model, then `prepare_model`.

What do you think?

Reply:
Got it. It's a bit embarrassing that in order to use "prepare_model" we are forced to do some weird things. This is because we fuse "move_to_gpu" and "wrap_with_ddp" together.

I'm wondering if we even need "prepare_model", because users should be able to wrap it with DDP/FSDP themselves. cc @matthewdeng

Reply:
Oh yes this is a good point. Another idea is to have a utility for checkpointing that will just checkpoint the CPU base model. This would allow us to use `prepare_model` + this utility in combination to support the (CPU, GPU) x (DDP, non-DDP) combination, but may not be fully efficient. (Actually, anyone know what Lightning does?) Users can then optimize the code with the PyTorch primitives.